src/Utilities/Options/EditorConfigParser.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Analyzer.Utilities
+{
+    /// <summary>
+    /// Parses a given .editorconfig source text into <see cref="CategorizedAnalyzerConfigOptions"/>.
+    /// </summary>
+    internal static class EditorConfigParser
+    {
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static readonly Regex s_propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        private static readonly StringComparer s_keyComparer = CaseInsensitiveComparison.Comparer;
+
+        /// <summary>
+        /// A set of keys that are reserved for special interpretation for the editorconfig specification.
+        /// All values corresponding to reserved keys in a (key,value) property pair are always lowercased
+        /// during parsing.
+        /// </summary>
+        /// <remarks>
+        /// This list was retrieved from https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties
+        /// at 2018-04-21 19:37:05Z. New keys may be added to this list in newer versions, but old ones will
+        /// not be removed.
+        /// </remarks>
+        private static readonly ImmutableHashSet<string> s_reservedKeys
+            = ImmutableHashSet.CreateRange(s_keyComparer, new[] {
+                "root",
+                "indent_style",
+                "indent_size",
+                "tab_width",
+                "end_of_line",
+                "charset",
+                "trim_trailing_whitespace",
+                "insert_final_newline",
+            });
+
+        /// <summary>
+        /// A set of values that are reserved for special use for the editorconfig specification
+        /// and will always be lower-cased by the parser.
+        /// </summary>
+        private static readonly ImmutableHashSet<string> s_reservedValues
+            = ImmutableHashSet.CreateRange(s_keyComparer, new[] { "unset" });
+
+        private static readonly ConditionalWeakTable<SourceText, CategorizedAnalyzerConfigOptions> s_cachedOptions
```

Comment:
I see so, you put the cache here.

Reply:
Yes, as `SourceText` is likely the entity shared across projects/snapshots. Additionally, I presume compiler will possibly cache based on the same key as well..

Reply:
I am not sure who is going to hold on to SourceText strongly. the http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,588 do that for ones given to TryGetValue, but otherwise, in IDE case, it might get kicked out from memory..

Reply:
if you are going to do this, doing it on AnalyzerOption might be better?

Reply:
Updated to now cache based on `AnalyzerOptions`. Verified that we only compute `CategorizedAnalyzerConfigOptions` once per compilation and editing either source or additional file leads to us recomputing the options.